src/libraries/System.Private.CoreLib/src/System/Char.cs

```diff
@@ -1049,5 +1059,741 @@ public static int ConvertToUtf32(string s, int index)
+            {
+                return (char)(decimal)(object)value;
+            }
+            else if (typeof(TOther) == typeof(double))
+            {
+                return (char)(double)(object)value;
+            }
+            else if (typeof(TOther) == typeof(short))
+            {
+                return (char)(short)(object)value;
+            }
+            else if (typeof(TOther) == typeof(int))
+            {
+                return (char)(int)(object)value;
+            }
+            else if (typeof(TOther) == typeof(long))
+            {
+                return (char)(long)(object)value;
+            }
+            else if (typeof(TOther) == typeof(nint))
+            {
+                return (char)(nint)(object)value;
+            }
+            else if (typeof(TOther) == typeof(sbyte))
+            {
+                return (char)(sbyte)(object)value;
+            }
+            else if (typeof(TOther) == typeof(float))
+            {
+                return (char)(float)(object)value;
+            }
+            else if (typeof(TOther) == typeof(ushort))
+            {
+                return (char)(ushort)(object)value;
+            }
+            else if (typeof(TOther) == typeof(uint))
+            {
+                return (char)(uint)(object)value;
+            }
+            else if (typeof(TOther) == typeof(ulong))
+            {
+                return (char)(ulong)(object)value;
+            }
+            else if (typeof(TOther) == typeof(nuint))
+            {
+                return (char)(nuint)(object)value;
+            }
+            else
+            {
+                ThrowHelper.ThrowNotSupportedException();
+                return default;
+            }
+        }
+
+        [RequiresPreviewFeatures]
+        static (char Quotient, char Remainder) INumber<char>.DivRem(char left, char right)
+            => ((char, char))Math.DivRem(left, right);
+
+        [RequiresPreviewFeatures]
+        static char INumber<char>.Max(char x, char y)
+            => (char)Math.Max(x, y);
+
+        [RequiresPreviewFeatures]
+        static char INumber<char>.Min(char x, char y)
+            => (char)Math.Min(x, y);
+
+        [RequiresPreviewFeatures]
+        static char INumber<char>.Parse(string s, NumberStyles style, IFormatProvider? provider)
+            => Parse(s);
+
+        [RequiresPreviewFeatures]
+        static char INumber<char>.Parse(ReadOnlySpan<char> s, NumberStyles style, IFormatProvider? provider)
+        {
+            if (s.Length != 1)
+            {
+                throw new FormatException(SR.Format_NeedSingleChar);
+            }
+            return s[0];
+        }
+
+        [RequiresPreviewFeatures]
+        static char INumber<char>.Sign(char value)
+            => (char)((value == 0) ? 0 : 1);
+
+        [RequiresPreviewFeatures]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static bool INumber<char>.TryCreate<TOther>(TOther value, out char result)
```

Comment:
I still question whether this is the right creation direction.  This requires a type T to know how to create itself from another type TOther, which means T needs to know about TOther.  That prevents, for example, Int64 being able to create itself from BigInteger.  If it were the other way around, BigInteger could know how to create an instance of any type it can see, so it could implement the cast to Int64.

Is that direction covered elsewhere?

Reply:
You result in the same fundamental issue in either direction. For an arbitrary, unknown `U`, you might not be able to create or convert to a given `T`.

There are some directions where going from `U` to `T` is easier and some where going from `T` to `U` will be easier. Right now the interfaces are only exposing "constructor" equivalents (create a `T` from a `U`, rather than convert a `U` to a `T`) because exposing both raises complications in determining the right way/order to cover both scenarios and avoid infinite recursion issues.

> That prevents, for example, Int64 being able to create itself from BigInteger

This is going to be solved by `IBinaryInteger.TryWriteBytes` and related methods. At some point, you have to be able to fallback to something "generic enough" so you don't need to codify "every possible type".

Having something like `IConvertibleTo<TOther>` would help, but its also not something that the core types can reasonably implement due to additional startup cost and number of conversions between all 13 primitive types.

Reply:
Hmmm, `TryWriteBytes` isn't in this PR, that isn't right 🤔

Reply:
> For an arbitrary, unknown U, you might not be able to create or convert to a given T.

No, but you have a much higher chance of doing so when you the thing with the data to be converted is able to reference the thing to be created.

Reply:
> No, but you have a much higher chance of doing so when the thing with the data to be converted is able to reference the thing to be created.

Can you elaborate, preferably with examples, of why `ConvertTo` has a higher chance of success than `CreateFrom`? My expectation is that you ultimately need both and that they should be roughly equivalent given internal implementation details.

The biggest problem is the pit of failure for having some `Create` method, without DIMs, that will correctly try both (to give the highest chance of success) without risking any kind of infinite recursion (or a potentially unbounded number of interfaces to check), etc.

Reply:
> Can you elaborate, preferably with examples, of why ConvertTo has a higher chance of success than CreateFrom?

Any type defined in any 3rd-party library can convert from its custom numeric type to Int32, Double, etc.  Int32, Double, etc. can only know about other types defined by us in Corelib, so they have no chance of converting from any such 3rd-party type.

Reply:
> so they have no chance of converting from any such 3rd-party type.

That's the scenario covered by `TryWriteBytes`. At the end of the day, you will have some list of hardcoded types and then a set of types that can be generically supported by understanding and getting the underlying representation (e.g. any `IBinaryInteger` or any `IFloatingPoint`, etc).

Outside of those, you run into the same problems with `ConvertFrom` and `ConvertTo` in that some type has to take a dependency on some other type and add it to the hardcoded list.

There are layering problems in both directions and both ultimately are needed. The problem that needs a solution is how to support both in a way that doesn't lead users into a pit of failure and that doesn't require explicitly implementing `IConvertibleTo<A>`, `IConvertibleTo<B>`, `IConvertibleTo<...>`

Reply:
> you run into the same problems with ConvertFrom and ConvertTo in that some type has to take a dependency on some other type and add it to the hardcoded list

Yes, but in one direction, every type in the world can see all of our important primitives, and in the other, our important primitives can't see anything else but themselves.  There's a huge difference.

Reply:
I was also wondering about this with my "what's the extensibility story here?" hat on. I'm not sure if it's realistic, but how would something like `NodaTime.Instant` be able to participate here?

Reply:
This all comes down to: "We need both in practice". There are flaws and limitations to just one or the other and so we ultimately need both `CreateFrom` and `ConvertTo` style APIs to ensure the relevant scenarios are covered.

@stephentoub and I did discuss this a bit last week and it will likely come down to some pattern where we tell users that you shouldn't have `CreateFrom` call `ConvertTo` or vice-versa, to avoid the issue with potential infinite recursion (imagine `CreateFrom<TOther>` falls back to calling `value.ConvertTo<TSelf>` for an unrecognized `TOther`, and `value.ConvertTo<TOther>` falls back to `TOther.CreateFrom<TSelf>`).

For implementors, there should likely be some recommended pattern to ensure that both functions cover all the relevant items. My initial thought is that we have:
* `static TSelf Create<TOther>(TOther value)`
* `static bool TryCreate<TOther>(TOther value, out TSelf result)`
* `static bool TryConvertFrom<TOther>(TOther value, out TSelf result)`
* `static bool TryConvertTo<TOther>(TSelf value, out TOther result)`

Where the standard pattern is something like:
```csharp
public static TSelf Create<TOther>(TOther value)
{
    if (!TryCreate(value, out TSelf result))
    {
        throw new NotSupportedException();
    }

    return result;
}

public static bool TryCreate<TOther>(TOther value, out TSelf result)
{
    return TryConvertFrom<TOther>(value, out result)
        || TryConvertTo<TSelf>(value, out result);
}

public static bool TryConvertFrom<TOther>(TOther value, out TSelf result)
{
    if (typeof(TOther) == typeof(...))
    {
        // ..., any TOther where ConvertFrom is most efficient
    }
    else
    {
        return false;
    }
}

public static bool TryConvertTo<TOther>(TSelf value, out TOther result)
{
    if (typeof(TOther) == typeof(...))
    {
        // ..., any TOther where ConvertTo is most efficient
    }
    else
    {
        return false;
    }
}
```

This would extend to each `Create*` pattern and so would include "exact", "saturating", "truncating", etc

Reply:
Do we need both Create and TryCreate variants, or could we just have the TryCreate?  Could Create be a DIM if/when we decide we need it?

Also, you talked about the fallback of being able to convert to/from bytes.  Where does that fit into this pattern?

Reply:
> Do we need both Create and TryCreate variants, or could we just have the TryCreate

We could probably have just `TryCreate` but that might make it less idiomatic to use in expressions. I don't have a strong preference here.

> For that matter could TryCreate be a DIM and you only implemenet TryConvertFrom/To?

That would be ideal, but is out of scope for .NET 6. It would be addressed before we ship for real based on whether or not DIMs are in the official release

> Also, you talked about the fallback of being able to convert to/from bytes. Where does that fit into this pattern?

You would handle it in one or both of the `TryConvert*` APIs as one of the `typeof() == typeof()` checks. Basically:
* Handle any concrete type
* Handle any abstract type via things like IBinaryInteger where you can get the raw 2s complement representation as bytes
* `return false`

This, for value type `TOther`, will compile down very efficiently. For reference types there will be casts and checks but still not bad overall.

Reply:
> We could probably have just TryCreate but that might make it less idiomatic to use in expressions

Fair enough.

What's your plan for factoring this all in?

Reply:
I'd like to get this PR merged so we have the base bits in and ready. Then to follow up with a PR that adds in `TryConvert*`, ideally before Preview 7, but possibly for RC1.